### DISCUSSION TITLE
Better tools for contribuiting

### DISCUSSION THREAD

**Discussion by feinstein** [firebase/flutterfire#3275] [General]
The Flutter repo has under the flutter tool `flutter update-packages` so all the nested folders with `pubspec.yaml` would have `pub get` called automatically. Also `flutter ide-config --overwrite` would set it up for IntelliJ IDEA.

Are there any tools for this repo as well? I can't find anything in the contributing docs. How do you guys do it? What's your workflow on this?

If there are tools, then the docs should describe them. If there aren't so maybe we should make them. Anything that helps on this matter is useful for contributing, because when I pull the git commits, that's how I get my IDE:

![image](https://user-images.githubusercontent.com/8120472/90856695-e8c3e180-e358-11ea-8520-4b3ca56a31fa.png)

The red folders means they are all ignored by the IDE.

If I mark them to not be ignored I get tons of errors:

![image](https://user-images.githubusercontent.com/8120472/90857423-79e78800-e35a-11ea-9cfc-10def558feda.png)

**Comment by Ehesp** [comment]
This reminds me, this PR needs to be updated and merged: https://github.com/FirebaseExtended/flutterfire/pull/2963

**Comment by feinstein** [comment]
If I follow the PR, will this errors disappear?

**Comment by Salakar** [comment]
> If I follow the PR, will this errors disappear?

They should do yes, though we had some issues earlier in dev with IntelliJ not playing nice with the number of packages so we all switched to Vscode and have no issues there:

![image](https://user-images.githubusercontent.com/5347038/90894007-da58e280-e3b7-11ea-89a3-280d2b75e3c8.png)

The tooling we created ([Melos](https://github.com/invertase/melos)) locally 'links' (amongst other features) all the dependent packages together based on their pubspec dependencies - previously if you made a change to say a platform interface, anything depending on that PI package wasn't aware of the change until it was published, which doesn't help developing. Now if you change that PI package, analyzer picks it up in all other packages that use it and highlights any issues. Can see in analyzer diagnostics that all packages get added as contexts now;

![image](https://user-images.githubusercontent.com/5347038/90894955-4720ac80-e3b9-11ea-8829-c502b08c14da.png)

**Comment by feinstein** [comment]
Is it my impression or this monorepo problem could be solved with a new feature into the Dart dependecy system? Like adding a new parameter to `pubspec.yaml`, letting the build system know there's a dependency for development only, probably a new behavior under `dev_dependencies`.

I am not experienced with monorepos, otherwise I would make a suggestion under the pub repo. Your close contact with google would sure help with this.

**Comment by feinstein** [comment]
Do any of you run `melos` under Windows 10?

I am getting this error:

![image](https://user-images.githubusercontent.com/8120472/90912379-e5f4db00-e3b0-11ea-8752-504643f1c8a0.png)

I tried to run it as Administrator and closed my IDE.

(maybe move this to the `melos` repository? I am not sure if it's a problem with `melos` or with the Contributing steps in the docs).

**Comment by Salakar** [comment]
Hey @feinstein - I will be shipping a new version of `melos` soon, with Windows support & also improved Intellij support;

![image](https://user-images.githubusercontent.com/5347038/91086385-c676e000-e646-11ea-85d4-13ed45f36d33.png)

**Comment by feinstein** [comment]
Ok, please let me know when you do, as that's the only thing stopping me from submitting PRs.

**Comment by Salakar** [comment]
Will be in the next hour or so, will ping here.

**Comment by feinstein** [comment]
Ok, should I run `pub upgrade melos` or something similar?

**Comment by feinstein** [comment]
@Salakar I executed `>pub global activate melos` to update it, then things appear to be working but I got the following warning while running `melos bootstrap`:

> Warning: skipped executing a script as "rm" is not supported on Windows.

I am not sure what that script was, but probably there are files being left behind that should have been cleaned by `rm`.

IntelliJ only has this warning:

![image](https://user-images.githubusercontent.com/8120472/91099917-90336380-e63a-11ea-8076-f7ebc8e42e83.png)

**Comment by Salakar** [comment]
In the process of cleaning up the repo #3326 - will merge that shortly, then you're good to go. Might want to `pub global activate melos` again also, pushed some quick patches

**Comment by Salakar** [comment]
Right, merged. Please try now

**Comment by feinstein** [comment]
Same warning about `rm`.

**Comment by Salakar** [comment]
That ones fine, ignore it - just for me really to know that I need to rework the scripts in https://github.com/FirebaseExtended/flutterfire/blob/master/melos.yaml to not use `rm` on Windows 

Everything should be working hopefully though.

**Comment by feinstein** [comment]
I think it is, any tests you want me to run?

**Comment by Salakar** [comment]
A good test is rename this method to anything or comment it out: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_core/firebase_core/lib/src/firebase.dart#L40 - Dart Analysis warnings should appear for a lot of packages that are relying on `firebase_core` - if it does then that means all the packages are being interlinked correctly.


Alt+Shift+F10 on IntelliJ on Windows should also bring up a run dialog with a bunch of scripts based on the ones defined in `melos.yaml`, could try the 'test' one: 
![image](https://user-images.githubusercontent.com/5347038/91102122-71e05e80-e660-11ea-949d-bea89360d29e.png)

**Comment by feinstein** [comment]
Commenting the code does bring up a bunch of errors, but `Melos Run -> 'test'` isn't as simple to run, it complains the test configuration is not setup correctly.

![image](https://user-images.githubusercontent.com/8120472/91105678-1786d400-e647-11ea-89fa-91a3bc3aec83.png)

If I select to run it anyway the terminal shows this error:

![image](https://user-images.githubusercontent.com/8120472/91105753-4d2bbd00-e647-11ea-91f8-cd1b9fc658c9.png)

**Comment by Salakar** [comment]
Oh interesting, on my windows box this is where the bin gets put, do you know where it is on yours? I think you can run `where melos` on cmd and it should print the location.

These scripts in the run configurations are just addional nices to haves, can also run them directly on the terminal using the `melos run` command for the time being until I can fix it. The main part was the package interlinking, so you're good to go 😅

**Comment by feinstein** [comment]
Mine's at: `C:\Users\Michel\AppData\Roaming\Pub\Cache\bin\melos.bat`. 

IIRC I just pointed my PATH to the Pub tool inside the Flutter installation folder, I didn't install Dart directly, maybe that's the source of the difference? I don't know. Maybe this can be automated with `where melos`?

**Comment by Salakar** [comment]
Great thanks for confirming. I just published a quick release with a temporary fix. Will look at parsing `where/which` output as a more robust fix later in the week, 1am here 😅

**Comment by feinstein** [comment]
Oh forgot you guys are in the UK, no problem, let me know if you want me to open an issue at the `melos` repo to track this.

**Comment by feinstein** [comment]
Just a FYI on the quick fix, the `postboostrap` option is broken.

![image](https://user-images.githubusercontent.com/8120472/91348156-c5b28b00-e7b9-11ea-87aa-5947beac28d3.png)

**Reply by Salakar** [reply]
Whats the issue with the script, pathing issue? I have a change pending to exclude `postX`/hook scripts from showing also.

**Reply by feinstein** [reply]
Yeah I think it's the path.

**Reply by Salakar** [reply]
Oh, ye I see the issue here, this is 'fixed' and it's no longer being generated anymore but the runConfig still exists in your .idea folder. You can go ahead and delete the file `.idea/runConfigurations/melos_run_postbootstrap.xml`

In a future release it will also cleanup any old/previously generated scripts that now no longer exist

**Reply by Salakar** [reply]
Can confirm by running bootstrap again after deleting the file, that file shouldn't get generated anymore

**Reply by feinstein** [reply]
Yep, I deleted it, then bootstraped again and there's no `postboostrap` option anymore.

**Comment by Salakar** [comment]
Forgot to mention, I pushed up a new build of melos late Monday; you should be seeing a whole lot more run configs, you can start typing when that dialog is up to easily filter, e.g. 'admob'. I've been using Android Studio now as my default editor for FlutterFire since Monday and everything seems to be ok.

![image](https://user-images.githubusercontent.com/5347038/91349773-ab8aa400-e7dd-11ea-9a5e-5fa8872d764a.png)

**Reply by feinstein** [reply]
I really liked this!

PS: The `postbootstrap` still has an error on this version.
